File: leetcode/01_99/83__Remove_Duplicates_from_Sorted_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }

 Ideas:
 - remove duplicated one by one
 - goto next non-duplciated one: delete multiple elements at a time
 */
class Solution1 {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode curNode = head.next, preNode = head;

        while(curNode != null) {
            if (curNode.val == preNode.val) {
                preNode.next = curNode.next;
            } else {
                preNode = curNode;
            }

            curNode = curNode.next;
        }

        return head;
    }
}

class Solution2 {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode cur = head;

        while(cur.next != null) {
            if (cur.val == cur.next.val) {
                cur.next = cur.next.next;
            } else {
                cur = cur.next; 
            }
        }

        return head;
    }
}

File: leetcode/01_99/83__Remove_Duplicates_from_Sorted_List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

typedef struct ListNode ListNode_t;

struct ListNode* deleteDuplicates(struct ListNode* head){
    if (head == NULL) {
        return NULL;
    }
    ListNode_t * cur = head;
    while(cur->next != NULL) {
        if (cur->val == cur->next->val) {
            cur->next = cur->next->next;
        } else {
            cur = cur->next;
        }
    }
    return head;
}

